rcl_lifecycle/src/rcl_lifecycle.c

```diff
@@ -0,0 +1,176 @@
+rcl_lifecycle_get_zero_initialized_state_machine()
+{
+  rcl_lifecycle_state_machine_t state_machine;
+  state_machine.transition_map.states = NULL;
+  state_machine.transition_map.states_size = 0;
+  state_machine.transition_map.transitions = NULL;
+  state_machine.transition_map.transitions_size = 0;
+  state_machine.com_interface = rcl_lifecycle_get_zero_initialized_com_interface();
+  return state_machine;
+}
+
+rcl_ret_t
+rcl_lifecycle_state_machine_init(
+  rcl_lifecycle_state_machine_t * state_machine,
+  rcl_node_t * node_handle,
+  const rosidl_message_type_support_t * ts_pub_notify,
+  const rosidl_service_type_support_t * ts_srv_change_state,
+  const rosidl_service_type_support_t * ts_srv_get_state,
+  const rosidl_service_type_support_t * ts_srv_get_available_states,
+  const rosidl_service_type_support_t * ts_srv_get_available_transitions,
+  bool default_states)
+{
+  if (rcl_lifecycle_com_interface_init(
+      &state_machine->com_interface, node_handle,
+      ts_pub_notify,
+      ts_srv_change_state, ts_srv_get_state,
+      ts_srv_get_available_states, ts_srv_get_available_transitions) != RCL_RET_OK)
+  {
+    return RCL_RET_ERROR;
+  }
+
+  if (default_states) {
+    rcl_lifecycle_init_default_state_machine(state_machine);
+  }
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_lifecycle_state_machine_fini(
+  rcl_lifecycle_state_machine_t * state_machine,
+  rcl_node_t * node_handle)
+{
+  rcl_ret_t fcn_ret = RCL_RET_OK;
+
+  fcn_ret = rcl_lifecycle_com_interface_fini(&state_machine->com_interface, node_handle);
+
+  rcl_lifecycle_transition_map_t * transition_map = &state_machine->transition_map;
+
+  for (unsigned int i = 0; i < transition_map->states_size; ++i) {
+    free(transition_map->states[i].valid_transition_keys);
+    free(transition_map->states[i].valid_transitions);
+  }
+  // free the primary states array
+  free(transition_map->states);
+  transition_map->states = NULL;
+  free(transition_map->transitions);
+  transition_map->transitions = NULL;
+
+  return fcn_ret;
+}
+
+rcl_ret_t
+rcl_lifecycle_state_machine_is_initialized(const rcl_lifecycle_state_machine_t * state_machine)
+{
+  if (!state_machine->com_interface.srv_get_state.impl) {
+    RCL_SET_ERROR_MSG("get_state service is null");
+    return RCL_RET_ERROR;
+  }
+  if (!state_machine->com_interface.srv_change_state.impl) {
+    RCL_SET_ERROR_MSG("change_state service is null");
+    return RCL_RET_ERROR;
+  }
+  return RCL_RET_OK;
+}
+
+const rcl_lifecycle_transition_t *
+rcl_lifecycle_is_valid_transition(
+  rcl_lifecycle_state_machine_t * state_machine,
+  rcl_lifecycle_ret_t key)
+{
+  unsigned int current_id = state_machine->current_state->id;
+  const rcl_lifecycle_state_t * current_state = rcl_lifecycle_get_state(
+    &state_machine->transition_map, current_id);
+
+  for (unsigned int i = 0; i < current_state->valid_transition_size; ++i) {
+    if (current_state->valid_transition_keys[i] == key) {
+      return &current_state->valid_transitions[i];
```

Comment:
Is this safe? What if valid_transitions is reallocated? Maybe valid transitions doesn't change?